caproto/ioc_examples/formula_ioc.py

```diff
@@ -0,0 +1,122 @@
+#!/use/bin/env python3
+from caproto.server import pvproperty, PVGroup, template_arg_parser, run
+from caproto import ChannelType
+import numpy as np
+from textwrap import dedent
+import ast
+
+np_names = set(np.__all__)
+np_ns = {k: getattr(np, k) for k in np_names}
+
+
+def extract_names(code_string):
+    output_name, _, expr = map(lambda x: x.strip(),
+                               code_string.partition('='))
+
+    ast_node = ast.parse(expr)
+    code_object = compile(expr, '', 'eval')
+    return (output_name,
+            expr,
+            code_object,
+            set(node.id for node in ast.walk(ast_node)
+                if (isinstance(node, ast.Name) and
+                    node.id not in np_names)))
+
+
+class FormulaIOC(PVGroup):
+    """
+    An IOC that computes a formula.
+
+    The equation is specified as a python assignment statement an IOC
+    boot.  It must be of the form::
+
+       ret_name = <expr>
+
+    The expression will be evaluated in a namespace with numpy's top-level
+    namespace available.
+
+    Any free-variables will be identified and used as the PV names of
+    the form `{prefix}{var}`.  The output will be published to
+    `{prefix}{ret_name}`
+
+
+    Readonly PVs
+    ------------
+    {prefix}formula -> the equation computed
+
+    {prefix}out_postfix -> the PV post-fix where the output is available
+    {prefix}vars -> the PV post-fixes where the inputs should be posted
+
+    {prefix}{ret_name} -> The result of the computation
+
+    Input PVs
+    ---------
+    {prefix}{var} -> the value for {var} in the inputs
+    """
+    proc = pvproperty(value=0)
+
+    def __init__(self, *args, code_object, var_names, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.code_object = code_object
+        self.var_names = var_names
+
+    @proc.putter
+    async def proc(self, instance, value):
+        ret = await self.compute(instance)
+        await self.output.write(ret)
+
+        return 0
```

Comment:
Should this catch any errors from compute and return 1?

Reply:
I don't think we have any sematics around return non-zero to mean failure on putters (as the return value of this is what is actually written to the channel).  If this raises, it will be caught by https://github.com/NSLS-II/caproto/blob/0f96128eb764054070b31e4a8c0e9b0e54407873/caproto/_data.py#L473-L486
and sets an alarm on proc.

We probably should make all of the PVs share the same alarm...

Reply:
They now all share an alarm (thanks @klauer for `alarm_group`!).